library/spdm_crypt_lib/fips/libspdm_selftest_eddsa.c

```diff
@@ -122,14 +105,6 @@ bool libspdm_fips_selftest_eddsa(void *fips_selftest_context)
         goto update;
     }
 
-    if (!libspdm_consttime_is_mem_equal(signature_25519, expected_signature_25519,
```

Comment:
why we cannot compare it?

Reply:
The signature isn't constant

Reply:
Good question! I have no idea. I was hoping whoever originally put the `evp_pkey_copy_downgraded()` call there would know.

For some reason not calling `evp_pkey_copy_downgraded()` means the signature is changing and I have no idea why. The [original commit](https://github.com/DMTF/libspdm/commit/882602c9487d1b15cbd892f2402b10640349b80f) added the call to  `evp_pkey_copy_downgraded()`, but I don't see a justification for the internal library call in the commit message

Reply:
@Wenxing-hou , please provide your comment.

Reply:
In the `libspdm_ecd_set_pri_key` and `libspdm_ecd_set_pub_key`, the purpose is setting the private key or public key to the `ecd_context`.

Using `libspdm_ecd_set_pub_key` as an example,  **the code in your PR is:**

```
    evp_key = (EVP_PKEY *)ecd_context;    (in line 102)

    evp_key = EVP_PKEY_new_raw_public_key(EVP_PKEY_id(evp_key), NULL,
                                          public_key, public_key_size);

    if (evp_key == NULL) {
        return false;
    }
```
This code will change the `evp_key` and will not achieve the goal(setting the private key or public key to the `ecd_context`).

**The origin code is:**
```
    evp_key = (EVP_PKEY *)ecd_context;  (in line 104)

    new_evp_key = EVP_PKEY_new_raw_public_key(EVP_PKEY_id(evp_key), NULL,
                                              public_key, public_key_size);

    if (new_evp_key == NULL) {
        return false;
    }

    if (evp_pkey_copy_downgraded(&evp_key, new_evp_key) != 1) {
        EVP_PKEY_free(new_evp_key);
        return false;
    }

    EVP_PKEY_free(new_evp_key);
```

Reply:
Then I find a API `EVP_PKEY_copy_parameters` in the `include\openssl\evp.h`.
**I changed the origin code:**
```
    new_evp_key = EVP_PKEY_new_raw_public_key(EVP_PKEY_id(evp_key), NULL,
                                              public_key, public_key_size);

    if (new_evp_key == NULL) {
        return false;
    }

    // if (evp_pkey_copy_downgraded(&evp_key, new_evp_key) != 1) {
    //     EVP_PKEY_free(new_evp_key);
    //     return false;
    // }

    if (EVP_PKEY_copy_parameters(evp_key, new_evp_key) != 1) {
        EVP_PKEY_free(new_evp_key);
        return false;
    }

    EVP_PKEY_free(new_evp_key);
    return true;
```

But the fips test will fail at:
```
    if (!libspdm_consttime_is_mem_equal(get_pub_key_25519, public_key_25519,
                                        sizeof(public_key_25519))) {
        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "EDDSA 25519 get key content compare failed \n"));
        libspdm_ecd_free(ecd_25519);
        result = false;
        goto update;
    }
```

Then I  searched for various materials. It seems that the `EVP_PKEY_copy_parameters` only copy part of the EVP_PKEY parameters.

Reply:
Yeah, `EVP_PKEY_copy_parameters()` isn't what you want.

It really seems like this should use `EC_KEY` instead of `EVP_PKEY`. The `ec.c` seems like a much better approach to do this

Reply:
Following: https://www.openssl.org/docs/manmaster/man7/Ed448.html and the OpenSSL code, `EVP_PKEY_set_octet_string_param()` should work, but it doesn't